### PR TITLE
Allow pure decay IndependentOperator

### LINKED ISSUE #2963: `IndependentOperator` does not work as intended for decay-only problem
## Bug Description
`IndependentOperator` claims that using an empty `MicroXS` object allows a decay-only step. It appears that trying to handle only decay, without any associated `openmc.Model` object, results in an error saying that no heat was found from the OpenMC output.

```
[openmc.deplete] t=0.0 s, dt=1.0 s, source=0.0
No energy reported from OpenMC tallies. Do your HDF5 files have heating data?
```

Well that's exactly what I expect! No transport-induced heat production, and pure decay. `IndependentOperator` suggests this is allowable use of it, to quote the documentation:

```
  Note that passing an empty :class:`~openmc.deplete.MicroXS` instance to the
  ``micro_xs`` argument allows a decay-only calculation to be run.
```

If I am not using this class correctly, the documentation should be improved.

## Steps to Reproduce

Here's an MWE that produces the error:

```
import numpy as np
import openmc
import openmc.deplete

chain_file = '/home/gavin/Data/chain_endfb71_pwr.xml'
openmc.config['chain_file'] = chain_file
openmc.config['cross_sections'] = '/home/gavin/Data/endfb-viii.0-hdf5/cross_sections.xml'

decay_times = np.array([1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0])

# Create one material to deplete
mat = openmc.Material()
mat.name = 'I decay!'
mat.add_nuclide('U239', 1.0, 'ao')
mat.volume = 10.0
mats = openmc.Materials([mat])

empty_micro = openmc.deplete.MicroXS(np.array([]).reshape((0, 0)), [], [])
op = openmc.deplete.IndependentOperator(mats, [1.0], [empty_micro], chain_file)
integ = openmc.deplete.PredictorIntegrator(op, decay_times, power=0.0, timestep_units='s')
integ.integrate()
```

I would hope to see the decay progeny of U239 appear here.

## Environment

On the latest commit 2974d53b3c07dc1a822f2b31ecd27b6553cbd458. Ubuntu. Nuclear data libraries are visible in the MWE.

### LINKED ISSUE #2968: Final step in depletion being written to depletion_results.h5 regardless of filename given to integrator
## Bug Description
The last step always tries to write to an h5 file called depletion_results.h5, even if a user gave something like my_decay.h5 to Integrator.integrate. This is because the kwarg for the file name is written at all steps but the last in the integrator base class.

It throws an exception as a result that looks like:

```
bigrig% python decay_mwe.py 
/usr/lib/python3/dist-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 1.26.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
[openmc.deplete] t=0.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=50.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=100.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=150.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=200.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=250.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=300.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=350.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=400.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=450.0 s, dt=50.0 s, source=0.0
[openmc.deplete] t=500.0 (final operator evaluation)
Traceback (most recent call last):
  File "/home/gavin/Scratch/decay_mwe.py", line 28, in <module>
    integ.integrate(path='decaystep.h5')
  File "/home/gavin/Code/openmc/openmc/deplete/abc.py", line 850, in integrate
    StepResult.save(self.operator, [n], res_list, [t, t],
  File "/home/gavin/Code/openmc/openmc/deplete/stepresult.py", line 570, in save
    results.export_to_hdf5(path, step_ind)
  File "/home/gavin/Code/openmc/openmc/deplete/stepresult.py", line 271, in export_to_hdf5
    res._to_hdf5(handle, step, parallel=False)
  File "/home/gavin/Code/openmc/openmc/deplete/stepresult.py", line 426, in _to_hdf5
    number_dset[index, i, low:high+1] = self.data[i]
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/home/gavin/.local/lib/python3.10/site-packages/h5py/_hl/dataset.py", line 997, in __setitem__
    mspace = h5s.create_simple(selection.expand_shape(mshape))
  File "/home/gavin/.local/lib/python3.10/site-packages/h5py/_hl/selections.py", line 264, in expand_shape
    raise TypeError("Can't broadcast %s -> %s" % (source_shape, self.array_shape))  # array shape
TypeError: Can't broadcast (1, 3820) -> (1, 3819)

```

## Steps to Reproduce

See the MWE in #2966 and set the path as follows:

```
integ.integrate(path='decaystep.h5')

results = openmc.deplete.Results('decaystep.h5')

```

## Environment

Does not matter.

### PR DESCRIPTION
**Pull request by gridley**
# Description

Fixes #2963 

Fixes #2968 

EDIT this also fixes #2968 with the second commit it adds.

Autopep8'd the independent operator file, and adds handling for decay only steps as we have in the coupled operator. This prevents an error where the heating is calculated as zero and an error is thrown erroneously. After taking this PR, the following code can produce this nice little plot in which we track the first two decay progeny of U239. It has a half life around 1400s and we can see that behavior nicely in this plot. Notably I need to turn off multiprocessing to have this work, described below.

It's so nice to have standalone decay like this, thanks @paulromano for implementing the `IndependentOperator` 🙂

```
import numpy as np
import openmc
import openmc.deplete
import matplotlib.pyplot as plt

# TODO I need help on why this does not work
openmc.deplete.USE_MULTIPROCESSING = False

chain_file = '/Users/gavin/Code/chain_endfb80_pwr.xml'
openmc.config['chain_file'] = chain_file
openmc.config['cross_sections'] = '/Users/gavin/Code/endfb-viii.0-hdf5/cross_sections.xml'

decay_times = np.ones(100) * 50.0

# Create one material to deplete. It is ten grams of U239.
mat = openmc.Material()
mat.name = 'I decay!'
mat.add_nuclide('U239', 1.0, 'ao')
mat.volume = 10.0
mat.set_density('g/cc', 1.0)

mats = openmc.Materials([mat])

empty_micro = openmc.deplete.MicroXS(np.array([]).reshape((0, 0)), [], [])
op = openmc.deplete.IndependentOperator(mats, [1.0], [empty_micro], chain_file)
integ = openmc.deplete.PredictorIntegrator(op, decay_times, power=0.0, timestep_units='s')
integ.integrate()

results = openmc.deplete.Results()

atoms_u239 = results.get_mass(mat, 'U239')
atoms_pu239 = results.get_mass(mat, 'Pu239')
atoms_np239 = results.get_mass(mat, 'Np239')
plt.semilogy(atoms_u239[0], atoms_u239[1])
plt.semilogy(atoms_np239[0], atoms_np239[1])
plt.semilogy(atoms_pu239[0], atoms_pu239[1])
plt.legend(('U239', 'Np239', 'Pu239'))
plt.xlabel('Time (s)')
plt.ylabel('Mass (g)')
plt.ylim([0.01, 11.0])
plt.show()

```

image

HOWEVER I am running into issues with multiprocessing set to true and may need some help. This error recurrently appears if depletion multiprocessing is turned on:

```
    super().__init__(process_obj)
  File "/opt/homebrew/Cellar/python@3.12/3.12.2_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/opt/homebrew/Cellar/python@3.12/3.12.2_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.2_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/spawn.py", line 164, in get_preparation_data
    _check_not_importing_main()
  File "/opt/homebrew/Cellar/python@3.12/3.12.2_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/spawn.py", line 140, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.

        To fix this issue, refer to the "Safe importing of main module"
        section in https://docs.python.org/3/library/multiprocessing.html
```

I guess we've missed some idiom here and haven't hit this race condition before because this case runs the depletion solver without much latency? If anyone knows where this `freeze_support` might reasonably go, I would very much appreciate your input.

# Checklist

- [x] I have performed a self-review of my own code
- [ ] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: openmc/deplete/abc.py [hunk 1]
```diff
@@ -848,7 +848,7 @@ def integrate(
                 print(f"[openmc.deplete] t={t} (final operator evaluation)")
             res_list = [self.operator(n, source_rate if final_step else 0.0)]
             StepResult.save(self.operator, [n], res_list, [t, t],
-                         source_rate, self._i_res + len(self), proc_time)
+                         source_rate, self._i_res + len(self), proc_time, path)
```
Comment (yardasol) [inline openmc/deplete/abc.py:851]: Nice catch!

**Review by yardasol — CHANGES_REQUESTED** [review]
Hi @gridley,

Good work on this quick fix. Fact-checking you here a bit, both @paulromano and myself are the authors of the `IndependentOperator` class :sweat_smile:. When we wrote `IndependentOperator` the intention was indeed to be able to run `IndependentOperator` in a decay-only mode. Checking through the regression and unit tests, it looks like we forgot to write one for this! Whoops! If you could add a regression test to this PR for this that'd be great!

**Comment by gridley** [comment]
Oh sick, thank you Olek! My apologies, unsure why I thought that. It is a lovely feature to have for analyzing reactors.

Will get the test in fosho.

**Comment by gridley** [comment]
OK @yardasol, I found that the test I've added here was able to reproduce the original problem, and passes with the introduction of these commits.

**Review by paulromano — APPROVED** [review]
Thanks for the fix @gridley!

**Comment by paulromano** [comment]
@yardasol are you satisfied with the changes in this PR?

**Review by yardasol — CHANGES_REQUESTED** [review]
Hi @gridley. Sorry for the delayed response. This is great. Can we also have a case where we have a non-empty `MicroXS` object that we run a depletion simulation on at zero power?

**Comment by gridley** [comment]
I would really prefer to not do that. The only potential problem with zero power is hitting that branch between calculating a normalizing factor and not, so I can't think of any difference that would take place with a filled MicroXS object.

**Review by yardasol — APPROVED** [review]

**Pull request merged**